sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -779,10 +804,11 @@ case class GpuCast(
 
       // convert dates that are in valid timestamp formats
       val converted =
-        convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_FULL, "%Y-%m-%dT%H:%M:%SZ%f",
-          convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_YYYY_MM_DD, "%Y-%m-%d",
-            convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_YYYY_MM, "%Y-%m",
-              convertTimestampOrNull(sanitizedInput, TIMESTAMP_REGEX_YYYY, "%Y"))))
+        convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_FULL_1, "%Y-%m-%d %H:%M:%S.%f",
+          convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_FULL_2, "%Y-%m-%dT%H:%M:%S.%f",
```

Comment:
Do we have to do an extra regex pass?  IIRC the regex passes are really expensive, and I'm curious how it compares to the cost of `isValidTimestamp` and `asTimestampMicrosceonds`.  I'm wondering if it would be overall cheaper to use the original regex and ifElse the two valid+conversion formats corresponding to that regex instead of adding another regex pass.

Reply:
Yes, that makes sense. I'll refactor this part to avoid the additional regex call.

Reply:
I pushed one commit that refactors this to use a single regex and then pushed another commit to remove the regex completely for this full timestamp case and tests pass with the current test values.

@revans2 would you have any concerns about removing regex use in this case?

Reply:
Once we have sanitized the inputs the regular expressions are redundant and should be removed.

Reply:
I think we can probably remove even more regular expressions from the code.

Reply:
I looked at this some more and remembered why we have the regex here. We need to make sure there is no additional text after the sanitized date/timestamp string. I added an additional test case to demonstrate this for timestamps and reverted the change to remove the regex for this case.

I do think could remove more of these regexes by checking the length of the sanitized strings (in conjunction with calling cuDF `isTimestamp`) rather than using regex and have filed an issue for this: https://github.com/NVIDIA/spark-rapids/issues/1738